Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::preprocess(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
+    collectLayersToFlattenRecursive(layersToFlatten);
+
+    for (auto* layer : layersToFlatten)
+        layer->flatten(options);
+}
+
+void TextureMapperLayer::collectLayersToFlattenRecursive(Vector<TextureMapperLayer*>& layers)
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Reset transformations for flattening root
```

Comment:
I see that, but why? :-) That should rather be explained.

Reply:
Updated comment